plugins/kotlin-dataframe/src/org/jetbrains/kotlinx/dataframe/plugin/impl/SimpleCol.kt

```diff
@@ -0,0 +1,127 @@
+package org.jetbrains.kotlinx.dataframe.plugin.impl
+
+import org.jetbrains.kotlin.fir.types.ConeKotlinType
+import org.jetbrains.kotlin.fir.types.ConeNullability
+import org.jetbrains.kotlin.fir.types.classId
+import org.jetbrains.kotlin.fir.types.isNullable
+import org.jetbrains.kotlinx.dataframe.impl.api.DataFrameLikeContainer
+import org.jetbrains.kotlinx.dataframe.impl.api.GenericColumn
+import org.jetbrains.kotlinx.dataframe.impl.api.GenericColumnGroup
+import org.jetbrains.kotlinx.dataframe.plugin.extensions.KotlinTypeFacade
+import org.jetbrains.kotlinx.dataframe.plugin.extensions.wrap
+import org.jetbrains.kotlinx.dataframe.plugin.impl.api.TypeApproximation
+import org.jetbrains.kotlinx.dataframe.plugin.pluginDataFrameSchema
+import org.jetbrains.kotlinx.dataframe.plugin.utils.Names
+
+data class PluginDataFrameSchema(
+    private val columns: List<SimpleCol>
+) : DataFrameLikeContainer<SimpleCol> {
+    override fun columns(): List<SimpleCol> {
+        return columns
+    }
+
+    override fun toString(): String {
+        return columns.asString()
+    }
+}
+
+private fun List<SimpleCol>.asString(indent: String = ""): String {
+    return joinToString("\n") {
+        val col = when (it) {
+            is SimpleFrameColumn -> {
+                "${it.name}*\n" + it.columns().asString("$indent   ")
+            }
+            is SimpleColumnGroup -> {
+                "${it.name}\n" + it.columns().asString("$indent   ")
+            }
+            is SimpleDataColumn -> {
+                "${it.name}: ${it.type}"
+            }
+        }
+        "$indent$col"
+    }
+}
+
+sealed interface SimpleCol : GenericColumn {
+    val name: String
+
+    override fun name(): String {
+        return name
+    }
+
+    fun rename(s: String): SimpleCol
+}
+
+data class SimpleDataColumn(
+    override val name: String,
+    val type: TypeApproximation
+) : GenericColumn, SimpleCol {
+
+    override fun name(): String {
+        return name
+    }
+
+    override fun rename(s: String): SimpleDataColumn {
+        return SimpleDataColumn(s, type)
+    }
+
+    fun changeType(type: TypeApproximation): SimpleDataColumn {
+        return SimpleDataColumn(name, type)
+    }
+
+}
+
+data class SimpleFrameColumn(
+    override val name: String,
+    private val columns: List<SimpleCol>
```

Comment:
If SimpleColumnGroup is a List<SimpleColumn>, then SimpleFrameColumn should be a List<List<SimpleColumn>> or a List<ColumnGroup> right?

Reply:
Ah no I see, it treats it like that because it tries to generalize the schema's, interesting